### PR TITLE
Add ImageNet-M evaluation task

### PR DESCRIPTION
**Pull request by vrv**
* Add ImageNet-M Evaluation example for paper from "When does dough become a bagel? Analyzing the remaining mistakes on ImageNet.".  Camera-ready for paper will be available soon (versions are also available on the web for those who want more details.)

* Tested with ResNet-18 evaluation and it reported 1 correct, which seems low, but it's possible my ImageNet validation data and torch setup is flawed.  Regardless, it does seem to run.

* To ensure that the evaluation code makes sense, this includes code that loads pre-computed Greedy Soups model logits and evaluates within evaluate(), verifying that evaluation code reproduces 19 correct predictions from paper.  I therefore have reasonably high confidence that most of the evaluation code is correct, and don't have enough knowledge of torch / imagenet cached predictions API to know whether that part is also correct.

# Task Description

## Dataset Creation
We select examples from the ImageNet multi-label set where 3 of 4 high performing models all make major mistakes.  We comprehensively label this small set of major mistakes, with the hope that future models should get close to 100% on, to demonstrate that a model generally avoids making obvious errors.

## Evaluation Metrics
Multi-label evaluation: ImageNet validation examples often have more than 1 correct label, so we allow the top prediction to match many correct labels.

## Expected Insights/Relevance
Given that ImageNet performance is close to saturation, we believe it is meaningful to assess whether top performing models are able to avoid making obvious mistakes on existing examples.

# Access

## Data
Data is simply a subset of the existing ImageNet validation set, with additional labels.  The json file referenced in this file has both the list of filenames comprising the subset, as well as their updated multi labels.

## License
CC-BY for our contributed metadata.

### CODE REVIEW

**Comment by vrv** [comment]
FYI I'm continuing to debug why ResNet-18 doesn't get great results, but if you have any other feedback, it's worth taking a look at now.

**Comment by vrv** [comment]
So I think the problem is:

model.imagenet_validation_result returns structures of length 50000 in an order that matches the Torch data loader.

We have new additional labels on our subset that we're loading in this file, and we index by filename.

We don't know how imagenet_validation_result order maps to filename order, so we can't map the results properly.

The ImageNet Torch DataLoader does not seem to have any indication of the order, so I can't even manually construct this mapping.

Any ideas on what to do here?

**Comment by vrv** [comment]
Okay, I found a way to use the Torch data loader to create the mapping from filename -> index, now I get

```
    Total ImageNet-M correct: 16
    ImageNet-M accuracy:  0.23529411764705882
```

as hoped!  Please take a look.

**Comment by vrv** [comment]
(I fixed the isort issue, please feel free to test again!)

File: examples/imagenet_m.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        filename = f"ILSVRC2012_val_{i+1:08}.JPEG".encode()
+        ret[filename] = logits_array[i]
+    return ret
+
+
+@sh_benchmark.register_task(
+    name="ImageNet-M", relative_data_folder="imagenet_m", standalone=True
+)
+@dataclasses.dataclass
+class ImageNetM(Task):
```
Comment (stes) [inline examples/imagenet_m.py:46]: Could you add a short docstring here?

**Review by stes — CHANGES_REQUESTED** [review]
Hi @vrv, thanks a lot for your contribution. We are working towards the first release of the benchmark and for this, released our [contribution guide](https://github.com/shift-happens-benchmark/icml-2022/blob/main/CONTRIBUTING.rst) with the aim of unifying the implementation of all contributions.

Please have a look, the required changes should be relatively small and are mostly concerns with placing the files within the package, and adding documentation.

From reviewing your PR, the main required changes are:

- [ ] Place the files into their own package inside the `tasks/` folder
- [ ] Add a README and optionally citation info
- [ ] We are also looking into the best ways to perform unit or integration tests and would be happy to get some feedback on how this would be most effective for you benchmark. Would it be possible to add a "dry-run/demo" option to the benchmark, and e.g. run it only on a small subset of the data? No immediate action/update is required right now, we are first gathering opinions from all authors on this.

Thanks a lot for your contribution, and please ping me if you have any questions. If you have technical questions, also feel free to ping @kotekjedi who is happy to help with implementation details.

**Comment by stes** [comment]
Hi!

This is an automated report for the documentation coverage of your PR. To provide good user experience for the benchmark, we aim to cover most functions with doc strings, using the [Google docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).

We would greatly appreciate your help in assuring that docstring coverage is sufficiently high for your contribution.

The current coverage is:
| Name                 |       Total |       Miss |       Cover |       Cover% |
|----------------------|-------------|------------|-------------|--------------|
| imagenet_m.py        |           4 |          2 |           2 |          50% |
|----------------------|-------------|------------|-------------|--------------|
| TOTAL                |           4 |          2 |           2 |        50.0% |
RESULT: FAILED (minimum: 80.0%, actual: 50.0%)

You can reproduce this result by running

```
interrogate -v --ignore-nested-functions --ignore-private --ignore-semiprivate --ignore-init-method
```

on the files you added to your PR.

File: examples/imagenet_m.py (hunk 1 above)
Comment (vrv) [reply examples/imagenet_m.py]: Done

File: shifthappens/tasks/__init__.py [hunk 2]
```diff
@@ -1 +1,3 @@
 from .base import Task  # noqa: F401
+
+from shifthappens.tasks import imagenet_m
```
Comment (kotekjedi) [inline shifthappens/tasks/__init__.py:3]:
```suggestion
from shifthappens.tasks import imagenet_m  # noqa: F401
```
Fix to pass mypy test.

File: shifthappens/tasks/imagenet_m/__init__.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+class ImageNetM(Task):
+    """Task definition for ImageNet-M, major mistakes eval subset.
+
+    This task evaluates a model on a small slice of the ImageNet
+    validation set.  The set has been additionally labeled with
+    multiple labels from the ImageNet label set by expert labelers.
+    The examples themselves are those where multiple high-performing
+    models in 2022 have yielded incorrect predictions where the
+    prediction was considered a "major" mistake.  
+
```
Comment (kotekjedi) [inline shifthappens/tasks/imagenet_m/__init__.py:55]:
```suggestion
    prediction was considered a "major" mistake.

```
Fix to pass black test.

File: shifthappens/tasks/imagenet_m/__init__.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    multiple labels from the ImageNet label set by expert labelers.
+    The examples themselves are those where multiple high-performing
+    models in 2022 have yielded incorrect predictions where the
+    prediction was considered a "major" mistake.  
+    
+    The goal of this evaluation task is to measure a model's ability
+    to avoid making obvious mistakes; in theory, models should be
+    capable of scoring near 100% on this evaluation, in addition to
+    high performance on other image classification benchmarks.
+    See the citation file for more information about how the dataset
```
Comment (kotekjedi) [inline shifthappens/tasks/imagenet_m/__init__.py:60]:
```suggestion
    See the citation file for more information about how the dataset
```
Fix to pass black test.

File: shifthappens/tasks/imagenet_m/__init__.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+    This task uses a multi-label evaluation approach: if the top
+    prediction from the model is found in the set of correct
+    or unclear labels for each example, the prediction is considered
+    correct.
+
+    The metric computes the total correct / percentage of correctly
+    classified multi-label examples.
+    """
+    def setup(self):
```
Comment (kotekjedi) [inline shifthappens/tasks/imagenet_m/__init__.py:71]:
```suggestion
    """
    
    def setup(self):
```
Fix to pass black test.

File: shifthappens/tasks/imagenet_m/__init__.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        if LOAD_GREEDYSOUPS_LOGITS:
+            self._greedy_soups = load_npy_as_dict(
+                os.path.join(dataset_folder, "greedysoups.npz")
+            )
+        else:
+            # Create mapping from torch DataLoader order to filename
+            self._filename_to_pred_index = {}
+            iif = tv_datasets.ImageFolder(root=imagenet.ImageNetValidationData)
+            for i, (filename, _) in enumerate(iif.samples):
+                im_filename = filename.split("/")[-1]
```
Comment (kotekjedi) [inline shifthappens/tasks/imagenet_m/__init__.py:149]:
```suggestion
                im_filename = os.path.normpath(os.path.basename(filename))
```
I believe it is better to use `os.path.basename` instead of `split`, to address the backslash in the path.

File: shifthappens/tasks/imagenet_m/__init__.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    def _evaluate(self, model: sh_models.Model) -> TaskResult:
+        """Computes multi-label accuracy on ImageNet-M slice."""
+        # Get imagenet validation results for the model.
+        imagenet_val_data = model.imagenet_validation_result
+
+        # For every image in imagenet_m, check whether prediction
+        # is in correct set for ImageNet-M.
+        correct = 0
+        for image_name, data in self._imagenet_m_2022.items():
+            image_id = int(image_name.split("_")[-1][:-5])
```
Comment (kotekjedi) [inline shifthappens/tasks/imagenet_m/__init__.py:161]: This variable is never used - should we remove it?

File: shifthappens/tasks/imagenet_m/__init__.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+                or predicted_class in data["unclear_multi_labels"]
+            ):
+                correct += 1
+
+        print("Total ImageNet-M correct:", correct)
+
+        return TaskResult(
+            accuracy=correct / len(self._imagenet_m_2022),
+            summary_metrics={Metric.Robustness: "accuracy"},
+        )
```
Comment (kotekjedi) [inline shifthappens/tasks/imagenet_m/__init__.py:180]:
```suggestion
        )
        
```

File: shifthappens/tasks/imagenet_m/__init__.py (hunk 7 above)
Comment (vrv) [reply shifthappens/tasks/imagenet_m/__init__.py]: Done.

**Review by vrv — COMMENTED** [review]
Moved.

**Comment by vrv** [comment]
Thanks for the review, please take a look

File: shifthappens/tasks/imagenet_m/__init__.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+    This task uses a multi-label evaluation approach: if the top
+    prediction from the model is found in the set of correct
+    or unclear labels for each example, the prediction is considered
+    correct.
+
+    The metric computes the total correct / percentage of correctly
+    classified multi-label examples.
+    """
+
```
Comment (kotekjedi) [inline shifthappens/tasks/imagenet_m/__init__.py:70]:
```suggestion

```

File: shifthappens/tasks/imagenet_m/__init__.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+            ):
+                correct += 1
+
+        print("Total ImageNet-M correct:", correct)
+
+        return TaskResult(
+            accuracy=correct / len(self._imagenet_m_2022),
+            summary_metrics={Metric.Robustness: "accuracy"},
+        )
+
```
Comment (kotekjedi) [inline shifthappens/tasks/imagenet_m/__init__.py:180]:
```suggestion

```

File: shifthappens/tasks/imagenet_m/__init__.py (hunk 9 above)
Comment (kotekjedi) [reply shifthappens/tasks/imagenet_m/__init__.py:70]: To pass black linter

File: shifthappens/tasks/imagenet_m/__init__.py (hunk 10 above)
Comment (kotekjedi) [reply shifthappens/tasks/imagenet_m/__init__.py:180]: To pass black linter.

File: shifthappens/tasks/imagenet_m/__init__.py [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+            self._imagenet_m_2022[image_name] = entry
+
+        if LOAD_GREEDYSOUPS_LOGITS:
+            self._greedy_soups = load_npy_as_dict(
+                os.path.join(dataset_folder, "greedysoups.npz")
+            )
+        else:
+            # Create mapping from torch DataLoader order to filename
+            self._filename_to_pred_index = {}
+            iif = tv_datasets.ImageFolder(root=imagenet.ImageNetValidationData)
```
Comment (kotekjedi) [inline shifthappens/tasks/imagenet_m/__init__.py:147]: Hi!
Could you please update your branch with the current state of the main, and replace `imagenet.ImageNetValidationData` with `shifthappens.config.imagenet_validation_path`?
